### PR TITLE
refactor: Extract `RegexDFAState` class, `RegexDFAStatePair` class, and `RegexDFAStateType` enum into their own files.

### PR DESCRIPTION
**Pull request by SharafMohamed**
- Doing a `git diff` will provide better comparisons for files with changed names than using the Github UI:
```
git fetch upstream
git fetch upstream pull/57/head:pr-57
git diff upstream/main pr-57 
```

# Description
Separate DFA Functionality into different files. 
- `RegexDFAState` class moved into its own file.
- `RegexDFAStatePair` class moved into its own file.
- `RegexDFAStateType` enum moved into its own file.

# Validation performed
Previously existing tests still succeed.

## Release Notes

- **New Features**
	- Introduced new header files for managing DFA states and their types.
	- Added functionality for creating and managing DFA states, including methods for transitions and acceptance checks.
	- Implemented logic for computing intersections of DFAs.

- **Bug Fixes**
	- Streamlined the DFA state management by consolidating redundant structures.

- **Documentation**
	- Updated documentation to reflect new classes and methods for better clarity on DFA state management.

### CODE REVIEW

File: src/log_surgeon/finite_automata/RegexDFAState.hpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    [[nodiscard]] auto next(uint32_t character) const -> RegexDFAState<stateType>*;
+
+private:
+    std::vector<uint32_t> m_matching_variable_ids;
+    RegexDFAState<stateType>* m_bytes_transition[cSizeOfByte];
+    // NOTE: We don't need m_tree_transitions for the `stateType ==
+    // RegexDFAStateType::Byte` case, so we use an empty class (`std::tuple<>`)
+    // in that case.
+    std::conditional_t<stateType == RegexDFAStateType::UTF8, Tree, std::tuple<>> m_tree_transitions;
+};
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/finite_automata/RegexDFAState.hpp]: Changed.

File: src/log_surgeon/finite_automata/RegexDFA.hpp [hunk 2]
```diff
@@ -142,8 +39,41 @@ class RegexDFA {
...
+        if (nfa_state->is_accepting()) {
+            dfa_state->add_matching_variable_id(nfa_state->get_matching_variable_id());
+        }
+    }
+    return dfa_state;
+}
+
+template <typename DFAStateType>
+auto RegexDFA<DFAStateType>::get_intersect(std::unique_ptr<RegexDFA> const& dfa_in
+) const -> std::set<uint32_t> {
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/finite_automata/RegexDFA.hpp:59]: Changed.

File: src/log_surgeon/finite_automata/RegexDFAState.hpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+template <RegexDFAStateType stateType>
+auto RegexDFAState<stateType>::next(uint32_t character) const -> RegexDFAState<stateType>* {
+    if constexpr (RegexDFAStateType::Byte == stateType) {
+        return m_bytes_transition[character];
+    } else {
```
[the remark this answers is not in the document]
Comment (SharafMohamed) [reply src/log_surgeon/finite_automata/RegexDFAState.hpp:63]: Resolved in future PR.

File: src/log_surgeon/finite_automata/RegexDFAState.hpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
+#ifndef LOG_SURGEON_FINITE_AUTOMATA_REGEX_DFA_STATE
```
Comment (LinZhihao-723) [inline src/log_surgeon/finite_automata/RegexDFAState.hpp:1]: Sorry to miss this in previous refactor PRs: I think we should name macros to exactly match the file name, so this header should be `LOG_SURGEON_FINITE_AUTOMATA_REGEXDFASTATE` instead. We can create an issue to keep track of this and fix them all together later

**Review by LinZhihao-723 — CHANGES_REQUESTED** [review]
The refactoring looks good to me. Let's make an agreement on the macro naming and then we can merge

File: src/log_surgeon/finite_automata/RegexDFAState.hpp (hunk 4 above)
Comment (SharafMohamed) [reply src/log_surgeon/finite_automata/RegexDFAState.hpp:1]: kk sounds good, I'll create the issue.

Comment (SharafMohamed) [reply src/log_surgeon/finite_automata/RegexDFAState.hpp:1]: Issue created.

**Review by LinZhihao-723 — APPROVED** [review]
The PR title looks good to me.
The macro naming issue is tracked here: https://github.com/y-scope/log-surgeon/issues/65

**Pull request merged**